### PR TITLE
Z3 solver

### PR DESCRIPTION
**Pull request by thewchan**
Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by thewchan** [comment]
@conda-forge/staged-recipes this successfully built for linux but not for osx and win; I can use some help here?

**Comment by ocefpaf** [comment]
For macOS you may need to do something [like this](https://github.com/libusb/libusb/issues/1097#issuecomment-1076736173).

**Comment by thewchan** [comment]
@ocefpaf I'll try that out, thanks! Any ideas about win build?

**Comment by thewchan** [comment]
@ocefpaf I'm getting a different OSX error now... this one is even more cryptic, do you have any ideas?

**Comment by ocefpaf** [comment]
Sorry, I'm terrible at macOS. Never had one. never even put my hands on one :-/

**Comment by thewchan** [comment]
@ocefpaf same tbh; trying some stuff now

**Comment by thewchan** [comment]
@ocefpaf I got OSX to work, but do you know anything about the win build? Still trying different things on my end

**Comment by carterbox** [comment]
Did you try looking at [what they do on their CI](https://github.com/Z3Prover/z3/blob/master/azure-pipelines.yml)? I find that is usually the quickest way to a working build.

**Comment by thewchan** [comment]
@carterbox hard to figure out what they did differently since I'm just running their build script as well... Do you have any pointers?

File: recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+--- setup.py	2022-07-06 15:24:04.000000000 -0400
++++ setup.py.new	2022-07-19 17:34:42.793030500 -0400
+@@ -123,7 +123,8 @@
+         'Z3_BUILD_DOCUMENTATION' : False,
+         'Z3_BUILD_TEST_EXECUTABLES' : False,
+         'Z3_BUILD_DOTNET_BINDINGS' : False,
```
[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: I tried that but maybe I'm doing it wrong? I'm doing it this way: https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2016%202019.html

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: Did you set the variable `CMAKE_GENERATOR`?

Sorry i didn't go through your whole history :/

[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: yeah I tried that as well, but I can try again, which do you think I should use?

[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: which generator, that is...

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: i can't really find the line in the source where things are failing. It is failing in some CMakeFile at line 5. But looking at hte source by hand, i can't find it.

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: sometimes.

[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: Okay I'm going to go ahead can try Ninja by setting env var just to see what gets spit out

[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: So specifically the error is:

```
  -- Building for: Ninja
  CMake Error at CMakeLists.txt:5 (project):
    Generator

      Ninja

    does not support platform specification, but platform

      x64

    was specified.


  CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage

```
The patch I made was to set the `CMAKE_CXX_COMPILER` option

[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/0002-CMAKE_CXX_COMPILER-win.patch:6]: But this doens't address the `done not support platform specification` error

File: recipes/z3-solver/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/z3-solver-{{ version }}.tar.gz
+  sha256: 5ed51c9748a23b6e21af9a526fa21f7ff0a76fd696aff436676300f162e10ddc
+  patches:
+    - 0001-target_os_osx.patch  # [osx]
+    # - 0002-CMAKE_CXX_COMPILER-win.patch  # [win]
+
+build:
+  script:
+    - set "CMAKE_GENERATOR=Ninja" # [win]
+    # - set 'CMAKE_GENERATOR_PLATFORM ="x64"'  # [win]
```
[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/meta.yaml:18]: yeah once everything is working I'll clean it up

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipes/z3-solver/meta.yaml:18]: i do'nt think it is getting commented out correctly. I don't always trust this selector stuff.

[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/meta.yaml:18]: The patch worked, the `CMAKE_CXX_COMPILER` error went away, but we're still getting:

```
  Configuring Z3
  -- Building for: Ninja
  CMake Error at CMakeLists.txt:5 (project):
    Generator

      Ninja

    does not support platform specification, but platform

      x64

    was specified.


  -- Configuring incomplete, errors occurred!
```

[the remark this answers is not in the document]
Comment (thewchan) [reply recipes/z3-solver/meta.yaml]: still getting the same error after removing the commented out code though

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipes/z3-solver/meta.yaml]: yeah. i'm just going a little crazy trying to find out where the platform specification is set.

File: recipes/z3-solver/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/z3-solver-{{ version }}.tar.gz
+  sha256: 5ed51c9748a23b6e21af9a526fa21f7ff0a76fd696aff436676300f162e10ddc
+  patches:
+    - 0001-target_os_osx.patch  # [osx]
+    - 0002-CMAKE_CXX_COMPILER-win.patch  # [win]
+
+build:
+  script:
+    - set "CMAKE_GENERATOR=Ninja"  # [win]
```
Comment (hmaarrfk) [inline recipes/z3-solver/meta.yaml:17]:
```suggestion
        - del /q /f pyproject.toml                       # [win]
        - set "CMAKE_GENERATOR=Ninja"  # [win]
```
Can we try this?

**Review by hmaarrfk — APPROVED** [review]

**Review by hmaarrfk — COMMENTED** [review]
did not mean to approve. wrong button before

File: recipes/z3-solver/meta.yaml (hunk 3 above)
Comment (hmaarrfk) [reply recipes/z3-solver/meta.yaml:17]: I'm somewhat worried that somebody is pulling in https://pypi.org/project/cmake/

**Comment by hmaarrfk** [comment]
:/ will have to take a look at later :(

**Comment by hmaarrfk** [comment]
My plan is to try to build locally, and have a look at that cmake log file. I'm just terrible at windows CMD systax and can't so it on the ci. 

Can you maybe make a bat file and the output the file they mention as part of that?

**Comment by thewchan** [comment]
> My plan is to try to build locally, and have a look at that cmake log file. I'm just terrible at windows CMD systax and can't so it on the ci.
> 
> Can you maybe make a bat file and the output the file they mention as part of that?

The problem is that I downloaded the sdist and there are many `CMakeList.txt` in the archive, but the error message didn't specify which one the error is from.

**Comment by hmaarrfk** [comment]
can you make a bld.bat file with:


```bat
@ECHO on
python -m pip install . -vv
type core/build/CMakeFiles/CMakeOutput.log
```

And comment out the line relating to the script.

**Comment by hmaarrfk** [comment]
can you make a bld.bat file with:


```bat
@ECHO on
python -m pip install . -vv
type core\build\CMakeFiles\CMakeOutput.log
```

And comment out the line relating to the script.

**Comment by thewchan** [comment]
> ```batchfile
> @ECHO on
> python -m pip install . -vv
> type core\build\CMakeFiles\CMakeOutput.log
> ```

Let me try that

**Comment by thewchan** [comment]
Something went awry here.... seems like it skipped building it all together

**Comment by hmaarrfk** [comment]
Make sure all the script sections are Linux only

**Comment by thewchan** [comment]
I might be printing out the wrong `CMakeList.txt`?

**Comment by hmaarrfk** [comment]
```
# This is the CMakeCache file.
# For build in directory: c:/Users/mark/mambaforge/conda-bld/debug_1658456019976/work/core/build
# It was generated by CMake: C:/Users/mark/mambaforge/conda-bld/debug_1658456019976/_build_env/Library/bin/cmake.exe
# You can edit this file to change values found and used by cmake.
# If you do not want to change any of the values, simply exit the editor.
# If you do want to change a value, simply edit, save, and exit the editor.
# The syntax for the file is as follows:
# KEY:TYPE=VALUE
# KEY is the name of a variable in the cache.
# TYPE is a hint to GUIs for the type of VALUE, DO NOT EDIT TYPE!.
# VALUE is the current value for the KEY.

########################
# EXTERNAL cache entries
########################

//Value Computed by CMake
CMAKE_PROJECT_DESCRIPTION:STATIC=

//Value Computed by CMake
CMAKE_PROJECT_HOMEPAGE_URL:STATIC=

//Value Computed by CMake
CMAKE_PROJECT_NAME:STATIC=Z3

//Value Computed by CMake
CMAKE_PROJECT_VERSION:STATIC=4.9.1.0

//Value Computed by CMake
CMAKE_PROJECT_VERSION_MAJOR:STATIC=4

//Value Computed by CMake
CMAKE_PROJECT_VERSION_MINOR:STATIC=9

//Value Computed by CMake
CMAKE_PROJECT_VERSION_PATCH:STATIC=1

//Value Computed by CMake
CMAKE_PROJECT_VERSION_TWEAK:STATIC=0

//Value Computed by CMake
Z3_BINARY_DIR:STATIC=C:/Users/mark/mambaforge/conda-bld/debug_1658456019976/work/core/build

//Value Computed by CMake
Z3_IS_TOP_LEVEL:STATIC=ON

//Value Computed by CMake
Z3_SOURCE_DIR:STATIC=C:/Users/mark/mambaforge/conda-bld/debug_1658456019976/work/core


########################
# INTERNAL cache entries
########################

//This is the directory where this CMakeCache.txt was created
CMAKE_CACHEFILE_DIR:INTERNAL=c:/Users/mark/mambaforge/conda-bld/debug_1658456019976/work/core/build
//Major version of cmake used to create the current loaded cache
CMAKE_CACHE_MAJOR_VERSION:INTERNAL=3
//Minor version of cmake used to create the current loaded cache
CMAKE_CACHE_MINOR_VERSION:INTERNAL=23
//Patch version of cmake used to create the current loaded cache
CMAKE_CACHE_PATCH_VERSION:INTERNAL=2
//Path to CMake executable.
CMAKE_COMMAND:INTERNAL=C:/Users/mark/mambaforge/conda-bld/debug_1658456019976/_build_env/Library/bin/cmake.exe
//Path to cpack program executable.
CMAKE_CPACK_COMMAND:INTERNAL=C:/Users/mark/mambaforge/conda-bld/debug_1658456019976/_build_env/Library/bin/cpack.exe
//Path to ctest program executable.
CMAKE_CTEST_COMMAND:INTERNAL=C:/Users/mark/mambaforge/conda-bld/debug_1658456019976/_build_env/Library/bin/ctest.exe
//Name of external makefile project generator.
CMAKE_EXTRA_GENERATOR:INTERNAL=
//Name of generator.
CMAKE_GENERATOR:INTERNAL=Ninja
//Generator instance identifier.
CMAKE_GENERATOR_INSTANCE:INTERNAL=
//Name of generator platform.
CMAKE_GENERATOR_PLATFORM:INTERNAL=x64
//Name of generator toolset.
CMAKE_GENERATOR_TOOLSET:INTERNAL=v142
//Source directory with the top level CMakeLists.txt file for this
// project
CMAKE_HOME_DIRECTORY:INTERNAL=C:/Users/mark/mambaforge/conda-bld/debug_1658456019976/work/core
//number of local generators
CMAKE_NUMBER_OF_MAKEFILES:INTERNAL=1
//Platform information initialized
CMAKE_PLATFORM_INFO_INITIALIZED:INTERNAL=1
//Path to CMake installation.
CMAKE_ROOT:INTERNAL=C:/Users/mark/mambaforge/conda-bld/debug_1658456019976/_build_env/Library/share/cmake-3.23


```

locally i get ^^^

**Comment by hmaarrfk** [comment]
I guess i don't know. i get slightly different errors locally. I would try maybe one mroe thing:

1. Building the C++ library on its own.

Then building the python wrappers around that.

**Comment by thewchan** [comment]
@conda-forge/core Does the core team know anyone who might be able to help with this one?

**Comment by jaimergp** [comment]
I am confused, where is that `setup.py` coming from? And why are we using PyPI for the source of a C++ package? Can we use the Github tarball + CMake/make directly?

https://github.com/Z3Prover/z3/blob/master/README-CMake.md

**Comment by thewchan** [comment]
@jaimergp The `setup.py` is in the sdist but also if you look at the github upstream, it's in the python api folder: https://github.com/Z3Prover/z3/tree/master/src/api/python

We are using PyPI to build this because this is building for the Python wrapper of the package, which some other package I'm trying to get on Conda Forge depends on.

**Comment by jaimergp** [comment]
Aaah, I see! I have taken a look at that setup.py and it looks overly complex :( I would be more comfortable doing the traditional CMake + make combo, which _should_ do the Python installation too, if I read their docs correctly. I don't have a Windows machine handy though...

**Comment by carterbox** [comment]
> @carterbox hard to figure out what they did differently since I'm just running their build script as well... Do you have any pointers?


> I am confused, where is that setup.py coming from? And why are we using PyPI for the source of a C++ package? Can we use the Github tarball + CMake/make directly?

Sorry for the slow reply. This is what I was trying to point out! Their CI builds are CMake driven, but this recipe is trying to drive the build by calling pip?

You should be splitting the CXX and Python into separate outputs/packages anyways because there is a public CXX API...

**Comment by thewchan** [comment]
In any case the osx and linux build is working; maybe that's the best we can do for now? @carterbox @jaimergp @hmaarrfk

**Comment by jaimergp** [comment]
If you don't need Windows for now, that's fine. Just add the `skip: True  # [win]` line under the `build` section.

File: recipes/z3-solver/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{% set name = "z3-solver" %}
+{% set version = "4.9.1.0" %}
```
Comment (hmaarrfk) [inline recipes/z3-solver/meta.yaml:2]: Seems like a new version just came out.

**Comment by hmaarrfk** [comment]
it seems that the same error is happening just building the C library
 https://github.com/conda-forge/staged-recipes/pull/19765

**Comment by thewchan** [comment]
@hmaarrfk thanks for trying! I might ask upstream for help at this point 🤷

**Comment by hmaarrfk** [comment]
I think after setting enough variable to "", I got the compilation going.
https://github.com/conda-forge/staged-recipes/pull/19765

I think it is your choice if you want to build the packages separately. Honestly, a 1 hour build time kinda sucks when you have to start building for python 3.7-3.10 even ignoring PyPy. Breaking up the C++ / Python package can help cut that quite a bit because, presumably, the python compilation will be quick.

I think you can get this recipe built with the same environment variables settings.

**Comment by thewchan** [comment]
@hmaarrfk Thanks so much for helping figuring this out! I'm up have doing it separately; I've not done that before so I'd need some guidance? Or if we can just use the one you made and rename this one `python-z3-solver` or something?

**Comment by hmaarrfk** [comment]
I don't really have any interest in maintaining this. So I would ask that you take on the burden of refactoring things to get them to work. it seems that you have set the correct cflags to get OSX to build correctly. 


It seems that they sometimes call the python internally `z3py`. I might follow that convention. or `z3` and `libz3`.

Generally speaking, you have to take on the work of "undoing" the bundling that people do for pypi. As others have said, the setup.py file looks rather overly complicated. Looking at the source, it seems that they are using ctypes to interface with the c-library

https://github.com/Z3Prover/z3/blob/master/src/api/python/z3/z3printer.py#L24=

This means that it is effectively a noarch package, and the setup.py file should be a boiler plate file.

The setup.py *should* look as simple as:

```python
from setuptools import setup
setup(
    name='z3-solver',
    version="4.10.1"
    description='an efficient SMT solver library',
    long_description='Z3 is a theorem prover from Microsoft Research with support for bitvectors, booleans, arrays, floating point numbers, strings, and other data types.\n\nFor documentation, please read http://z3prover.github.io/api/html/z3.html\n\nIn the event of technical difficulties related to configuration, compilation, or installation, please submit issues to https://github.com/z3prover/z3.git',
    author="The Z3 Theorem Prover Project",
    maintainer="Audrey Dutcher and Nikolaj Bjorner",
    maintainer_email="audrey@rhelmot.io",
    url='https://github.com/Z3Prover/z3',
    license='MIT License',
    keywords=['z3', 'smt', 'sat', 'prover', 'theorem'],
    packages=['z3'],
)
```

Goodluck

**Comment by thewchan** [comment]
@hmaarrfk thanks in any cases; if you can link me here the settings that worked (which var did you set?) that would be great!

**Comment by thewchan** [comment]
@hmaarrfk @conda-forge/core are there any example repo that did something similar that I can study on?

**Comment by hmaarrfk** [comment]
https://github.com/conda-forge/staged-recipes/pull/19765/files#diff-d7b5a2cdcba9ae8ab7eaff1f280a06aa8f3b305528e49e0922b93549bb34b5c7R10

**Comment by thewchan** [comment]
@hmaarrfk thanks again!

**Comment by jaimergp** [comment]
I searched cs.github.com with `org:conda-forge path:meta.yaml outputs pin_compatible python` and found a couple examples, but there are more.

I think [this one](https://github.com/conda-forge/lalpulsar-feedstock/tree/main/recipe) is similar to what you need to do.

**Pull request closed without merging**